src/test/java/software/amazon/awssdk/crt/test/Mqtt5ClientTestFixture.java

```diff
@@ -144,25 +147,38 @@ public void onMessageReceived(Mqtt5Client client, PublishReturn result) {
     }
 
     static final class PublishEvents_Futured_Counted implements PublishEvents {
+        static final ScheduledExecutorService TIMEOUT_SCHEDULER = Executors.newScheduledThreadPool(0);
+
+        // The "main" future which is intended to be used for waiting for all publishes.
         CompletableFuture<Void> publishReceivedFuture = new CompletableFuture<>();
+        // Additional future which helps with ensuring that no publishes beyond expected were received.
+        CompletableFuture<Void> afterCompletionFuture = new CompletableFuture<>();
+
         int currentPublishCount = 0;
         int desiredPublishCount = 0;
+        int afterCompletionWaitSec = 1;
         List<PublishPacket> publishPacketsReceived = new ArrayList<PublishPacket>();
+        HashMap<Mqtt5Client, Integer> clientsReceived = new HashMap<Mqtt5Client, Integer>();
 
         @Override
         public void onMessageReceived(Mqtt5Client client, PublishReturn result) {
             currentPublishCount += 1;
             if (currentPublishCount == desiredPublishCount) {
                 publishReceivedFuture.complete(null);
+                // An optional wait to ensure that no packets beyond expectations are arrived.
+                TIMEOUT_SCHEDULER.schedule(
+                        () -> afterCompletionFuture.complete(null), afterCompletionWaitSec, TimeUnit.SECONDS);
             } else if (currentPublishCount > desiredPublishCount) {
-                publishReceivedFuture.completeExceptionally(new Throwable("Too many publish packets received"));
+                afterCompletionFuture.completeExceptionally(new Throwable("Too many publish packets received"));
```

Comment:
This block did not work as intended prior to this PR, as `publishReceivedFuture` is already completed at this point (unless `desiredPublishCount == 0`, but it's not the case here).
I decided to fix it, though maybe it's better to remove this check for good, because this class is used in two tests: "Happy path. No drop in connection, no retry, no reconnect" and "Shared subscriptions", where this functionality is extra.